### PR TITLE
Fix [LB-41] - Upload lastfm backup

### PR DESCRIPTION
**Pull request by pinkeshbadjatiya**
Fix  	<a href="http://tickets.musicbrainz.org/browse/LB-41">#LB-41</a>
- Upload lastfm backup `json` file
- Supports uploading **.ZIP** files
- ZIP has to be in the same format as downloaded from lastfm.
- Sample file for testing - 
[21192540-QD3jC21IfeW40cZFN7AJ5AY7upcq49FR.zip](https://github.com/metabrainz/listenbrainz-server/files/294122/21192540-QD3jC21IfeW40cZFN7AJ5AY7upcq49FR.zip)

- Screenshots
Screenshot
<hr/>
image
<hr/>
Screenshot

### CODE REVIEW

**Comment by pinkeshbadjatiya** [comment]
Ready for review :)

**Comment by pinkeshbadjatiya** [comment]
Ready for review :)

File: webserver/views/user.py [hunk 1]
```diff
@@ -98,6 +99,63 @@ def import_data():
...
 
+@user_bp.route("/upload", methods=['POST'])
+@login_required
+def upload():
+    if request.method == 'POST':
+        f = request.files['file']
+        if f:
+            augmented_listens = []
+            msb_listens = []
+            try:
```
Comment (alastair) [inline webserver/views/user.py]: The data file that I sent you isn't actually the whole archive file. It is in fact a zip, containing many csvs and many json files. I've attached an example.
[21192540-QD3jC21IfeW40cZFN7AJ5AY7upcq49FR.zip](https://github.com/metabrainz/listenbrainz-server/files/204213/21192540-QD3jC21IfeW40cZFN7AJ5AY7upcq49FR.zip)
We should import the json data.

File: webserver/views/user.py [hunk 2]
```diff
@@ -98,6 +99,63 @@ def import_data():
...
+                    msb_listens = []
+
+            if msb_listens:
+                augmented_listens.extend(_messybrainz_lookup(msb_listens))
+
+            _send_listens_to_kafka("import", augmented_listens)
+    return render_template('user/import.html', user=current_user)
+
+
+def _convert_to_native_format(data):
```
Comment (alastair) [inline webserver/views/user.py]: This is a great example of a function which should be tested!

File: webserver/views/user.py [hunk 3]
```diff
@@ -98,6 +99,63 @@ def import_data():
...
+                msb_listens.append(dic)
+                if len(msb_listens) >= MAX_ITEMS_PER_MESSYBRAINZ_LOOKUP:
+                    augmented_listens.extend(_messybrainz_lookup(msb_listens))
+                    msb_listens = []
+
+            if msb_listens:
+                augmented_listens.extend(_messybrainz_lookup(msb_listens))
+
+            _send_listens_to_kafka("import", augmented_listens)
+    return render_template('user/import.html', user=current_user)
```
Comment (alastair) [inline webserver/views/user.py]: When this page is rendered on a `POST` we should include a message telling the user that their data was imported.

File: webserver/views/user.py [hunk 4]
```diff
@@ -98,6 +99,63 @@ def import_data():
...
+            try:
+                jsonlist = json.load(f)
+                if not isinstance(jsonlist, list):
+                    raise ValueError
+            except ValueError:
+                raise BadRequest("Not a valid lastfmbackup file.")
+
+            for listen in jsonlist:
+                dic = _convert_to_native_format(listen)
+                _validate_listen(dic)
```
Comment (alastair) [inline webserver/views/user.py]: This code is copied from the `submit` method in the API view. We shouldn't be duplicating this code. Instead, it should be put into a utilities module, which upload and submit can both call. In addition, making it a utility method makes it much easier to test.

File: webserver/views/user.py (hunk 3 above)
Comment (pinkeshbadjatiya) [inline webserver/views/user.py]: Added in commit  2834c7a

File: webserver/views/user.py (hunk 1 above)
Comment (pinkeshbadjatiya) [inline webserver/views/user.py]: I tested with all the json files and i works fine. :)

File: webserver/views/user.py (hunk 4 above)
Comment (pinkeshbadjatiya) [inline webserver/views/user.py]: Updated in commit  	ca8d00b

File: webserver/views/user.py [hunk 5]
```diff
@@ -127,7 +127,8 @@ def upload():
                 augmented_listens.extend(_messybrainz_lookup(msb_listens))
 
             _send_listens_to_kafka("import", augmented_listens)
-    return render_template('user/import.html', user=current_user)
+            flash('Congratulations !! Your listens have been uploaded successfully.')
```
Comment (gentlecat) [inline webserver/views/user.py]: `Congratulations! Your listens have been uploaded successfully.`

File: webserver/templates/user/import.html [hunk 6]
```diff
@@ -78,6 +78,13 @@ <h2 class="page-title">{{ user.musicbrainz_id }}</h2>
         {{ user.auth_token }}
      </div>
 
+     <p><b> or </b></p>
+     <p>Import your listens from the lastfm backup file.</p>
+     <div>
+         <form action="upload" method="POST" enctype=multipart/form-data>
+             <input type=file name=file>
+             <input type=submit value=Upload>
+         </form>
```
Comment (gentlecat) [inline webserver/templates/user/import.html]: It's usually a good idea to use Flask-WTF for creating forms.

**Comment by alastair** [comment]
In principle this seems good. I'll review the code again and merge if it's good.
We should put a flag in front of the option so that we can disable it in a hurry if we find that we can't process large files which people send us.

File: webserver/templates/user/import.html [hunk 7]
```diff
@@ -55,6 +55,13 @@ <h2 class="page-title">{{ user.musicbrainz_id }}</h2>
         {{ user.auth_token }}
      </div>
 
+     <p><b> or </b></p>
+     <p>Import your listens from the lastfm backup file.</p>
```
Comment (mayhem) [inline webserver/templates/user/import.html]: Please change the text to:

"Import your listens from a Last.fm backup file (.zip) or from a ListenBrainz export file (.json):"

And make it visually look more like the two sections above.

File: webserver/views/user.py [hunk 8]
```diff
@@ -138,6 +142,57 @@ def export_data():
...
+        # Check upload folder
+        if not 'UPLOAD_FOLDER' in current_app.config:
+            raise InternalServerError("Could not upload the file. Upload folder not specified")
+        upload_path = path.join(path.abspath(current_app.config['UPLOAD_FOLDER']), current_user.musicbrainz_id)
+        if not path.isdir(upload_path):
+            makedirs(upload_path)
+
+        # Write to a file
+        filename = path.join(upload_path, secure_filename(f.filename))
+        f.save(filename)
```
Comment (mayhem) [inline webserver/views/user.py]: This file is never removed when you're done with it.

File: config.py.sample [hunk 9]
```diff
@@ -53,3 +53,11 @@ LASTFM_API_KEY = "USE_LASTFM_API_KEY"
...
 # Set to True if Less should be compiled in browser. Set to False if styling is pre-compiled.
 COMPILE_LESS = True
+
+# MAX file size to be allowed for the lastfm-backup import, default is infinite
+# Size is in bytes
+MAX_CONTENT_LENGTH = 16 * 1024 * 1024  # 16MB
+
+# Specify the upload folder where all the lastfm-backup will be stored
+# The path must be absolute path
+UPLOAD_FOLDER = "METABRAINZ/listenbrainz-server/lastfm-backup-upload"
```
Comment (mayhem) [inline config.py.sample]: Lets use "/tmp/lastfm-backup-upload" as the default here.

File: .gitignore [hunk 10]
```diff
@@ -4,3 +4,6 @@
 *.pyc
 config.py
 *.log
+.cache
+env
+lastfm-backup-upload
```
Comment (mayhem) [inline .gitignore]: If we use the /tmp/lastfm-backup-upload default, then this doesn't need to be here.

**Pull request merged**